domain/network/state/linklayer_merge.go

```
@@ -229,22 +211,141 @@ func (st *State) applyMergeLinkLayerChanges(
 	return nil
 }
 
+// updateAddressSubnet updates the subnet_uuid field in the ip_address table.
+func (st *State) updateAddressSubnet(
+	ctx context.Context, tx *sqlair.TX,
+	addressUUID, subnetUUID string,
+) error {
+	type params struct {
+		SubnetUUID  string `db:"subnet_uuid"`
+		AddressUUID string `db:"address_uuid"`
+	}
+	stmt, err := st.Prepare(`
+UPDATE ip_address
+SET subnet_uuid = $params.subnet_uuid
+WHERE uuid = $params.address_uuid
+`, params{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+	return tx.Query(ctx, stmt, params{
+		SubnetUUID:  subnetUUID,
+		AddressUUID: addressUUID,
+	}).Run()
+}
+
+func (st *State) cleanupUniqueAddressOrphanSubnets(ctx context.Context, tx *sqlair.TX) error {
+	type orphan struct {
+		UUID string `db:"uuid"`
+	}
+	// Fetch orphan subnets
+	stmt, err := st.Prepare(`
+SELECT s.uuid as &orphan.uuid
+FROM subnet as s
+LEFT JOIN ip_address as a ON s.uuid = a.subnet_uuid
+LEFT JOIN provider_subnet as ps ON s.uuid = ps.subnet_uuid
+WHERE a.uuid IS NULL -- orphan subnet, linked to no addresses
+AND ps.provider_id IS NULL -- subnet without any provider id
+AND (
+    s.cidr LIKE '%.%/32' -- single address ipv4 subnet 
+    OR  
+    s.cidr LIKE '%:%/128' -- single address ipv6 subnet 
+    )`, orphan{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+	var orphanSubnets []orphan
+	if err := tx.Query(ctx, stmt).GetAll(&orphanSubnets); err != nil &&
+		!errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Errorf("getting orphan subnets: %w", err)
+	}
+	if len(orphanSubnets) == 0 {
+		return nil
+	}
+
+	// remove orphan subnets
+	return st.removeSubnets(ctx, tx, transform.Slice(orphanSubnets, func(o orphan) string { return o.UUID }))
+}
+
+// removeSubnets removes subnets from the subnet table.
+func (st *State) removeSubnets(
+	ctx context.Context, tx *sqlair.TX,
+	subnetUUIDs []string,
+) error {
+	// First remove any provider subnet mappings
+	type uuids []string
+	stmt, err := st.Prepare(`
+DELETE FROM provider_subnet
```

Don't remove these. Let the removal fail if the subnet has any associated provider IDs. We will know that these added by us as a fallback, rather by provider network discovery.